### DISCUSSION TITLE
How to configure symfony/mailer without DSN?

### DISCUSSION THREAD

**Discussion by crtl** [symfony/symfony#51253] [Q&A]
With swiftmailer-bundle it was possible to just pass username, password and other SMTP configuration in the swiftmailer.yaml config file.
With symfony/mailer it seems this option was removed completely or I cant find it.
Is it possible to configure mailer without DSN or do I have to use DSN?
Why was username/password manual config removed?

What is the purpose of DSN? Already with swiftmailer it felt very cumbersome to use because of all the URL char restrictions (it is very common to have special chars in passwords).
The documentation says to use `urlencode` for my credentials but why bother using env vars when I have to hardcode my credentials in source anyways? I dont get it.

**Comment by stof** [comment]
The use of the DSN was chosen because it allows to embed all the info in a single parameter. And also because this is the way most PaaS expose credentials so they can be used out of the box.

**Comment by crtl** [comment]
I generally always had the problem that in my prod environment the credentials are exposed as env variables so I cant control that they are encoded. It would be nice to be able to just  pass username, password etc..

For now I have just created my own env-var processor which pretty much fixes the issue:

````php
<?php

namespace App\DependencyInjection;

use Symfony\Component\DependencyInjection\EnvVarProcessorInterface;

/**
 * Adds urlencode processor for %env()% function in config.
 * Usage: %env(urlencode:MY_ENV_VAR)%
 */
class UrlencodeEnvVarProcessor implements EnvVarProcessorInterface
{

    public function getEnv(string $prefix, string $name, \Closure $getEnv)
    {
        $env = $getEnv($name);

        return urlencode($env);
    }

    public static function getProvidedTypes(): array
    {
        return [
            "urlencode" => "string"
        ];
    }
}

**Comment by stof** [comment]
Changing factories to not be based on DSN would be a big BC break and would make the integration harder in FrameworkBundle as well. So I would vote against it.

However, I strongly encourage you to send a PR adding the `urlencode` env processor in core for Symfony 6.4. This looks like a nice addition for cases like yours where you have separate variables.

**Reply by crtl** [reply]
I have never contributed here before but i will give it a try.

**Comment by tourze** [comment]
You can create you own transport by \Symfony\Component\Mailer\Transport::fromString
